### ISSUE TITLE
Spike: review EasyAuth references provided by FastTrack

### ISSUE THREAD

**Issue by ryanhofdotgov** [18F/piipan#1096]
See [Teams thread ](https://teams.microsoft.com/l/message/19:5222bb7a1faa47daa8b5f49e1d9b0d81@thread.tacv2/1623351360258?tenantId=ed5b36e7-01ee-4ebc-867e-e03cfa0d4697&groupId=be37a6c1-d322-49a6-805b-dc7d50d223c2&parentMessageId=1623351360258) for EasyAuth references provided by FastTrack. In particular, looking to understand how we can configure so as to drop session due to inactivity.

Also looking to understand what establishes the cookie lifetimes, etc. at the App Service level.

**Comment by timoballard** [comment]
Cookie lifetime can be configured via the `authsettingsv2` ARM template (`properties.login.cookieExpiration.timeToExpiration`). I tested this in the TTS sandbox and verified that the cookie expires after the configured timespan. When the user attempts to take action after this time, the backend attempts to refresh the session with the identity provider. If the IDP session is still active, a new cookie is issued. If the IDP session has expired, the user is redirected to the IDP login page. 

Configuration changes for 15 minute expiration can be seen [here](https://github.com/18F/piipan/commit/9e3c1b32e43ad23667a39cad7949d35ce61b5975).

IDP session duration can be configured for the TTS sandbox in the portal by going to:
`TTS-B2C-DEV -> Azure AD B2C -> User flows -> B2C_1_si -> Properties -> Session behavior`

cc: @ryanhofdotgov

**Comment by ryanhofdotgov** [comment]
@timoballard Nice – how did you figure out the required format for timeToExpiration? Did one of those links we get from FastTrack document that? [NIST will require this to be at most 12 hours](https://pages.nist.gov/800-63-3/sp800-63b.html), and we may not need to tinker with it, if we can just inherit from the IdP (`IdentityProviderDerived`, which I'm guessing is the default?) and we can trust the partner agency IdP is already doing the right thing. It may be really short like 10 minutes, or may be a full work day or more.

Did you find anything that might support in expiring the app session after inactivity? The configuration example above will always expire the app session at `t0`+15 minutes, even if at `t0`+14 minutes the user was actively using the application, where `t0` is the time of the initial OIDC authn. 
1. `t0` - login
1. `t0` + 1 minute - click a web page button, still works
1. `t0` + 14 minutes - click a web page button, still works
1. `t0` + 16 minutes - click a web page button, still works

instead of

1. `t0` - login
1. `t0` + 1 minute - click a web page button, still works
1. `t0` + 16 minutes - click a web page button, session expired due to inactivity, redirected to IdP

[Here's some corresponding NIST guidance](https://pages.nist.gov/800-63-FAQ/#q-b09). The intent gets a little muddied if you just get re-auth'd automatically when you get redirected to the IdP, but you can hypothetically control that at the IdP level.

**Comment by timoballard** [comment]
@ryanhofdotgov - Based on my testing, the `FixedTime` cookie expiration means that the web app doesn't check back in with the idP until the expiration has passed, regardless of user activity. There's a configurable session timeout on the B2C side as well, which can be marked as `Rolling` or `Absolute`.

So with `cookieExpiration` set to `FixedTime`, 5 minutes and the idP session timeout set to `Rolling`, 15 minutes, you'll get:

1. `t0` - login
2. `t0` + 4 minutes - click on a web page button, still works
3. `t0` + 6 minutes - click on a web page button, cookie expired - round trip to idP for new auth code, backend does token exchange (no user interaction), new cookie issued
4. `t0` + 22 minutes - click on a web page button, cookie expired - redirect to idP, user logs in for new auth code, backend does token exchange, new cookie issued

Setting the idP session timeout to `Absolute` does what you'd expect - the token exchange at `t0`+6 doesn't reset the idP session clock, so a web page action at `t0`+16 would require the user to log in at the idP.

I thought that using `IdentityProviderDerived` would mean that `cookieExpiration` is automatically set to match the token lifetime from the idP, but that doesn't seem to be the case. When I use `IdentityProviderDerived`, the cookie stays good well beyond the configured session timeout on the idP side. I read somewhere that the default cookie expiration is 8 hours, so I'm guessing it's falling back to that in this case.

**Comment by ryanhofdotgov** [comment]
@timoballard OK – so it looks like when we next sync up with FastTrack on this, we can ask what `IdentityProviderDerived` means, as your experimentation doesn't seem to hold up with most natural interpretation.

We can close this research spike – but still curious – how did you figure out the required format for timeToExpiration?

**Comment by timoballard** [comment]
@ryanhofdotgov - Ah - yes, forgot about that part of your question:

Initially I assumed that the `timeToExpiration` value was numeric, but wasn't sure if it was seconds, minutes, etc. Putting in numeric values didn't seem to work as expected.

Eventually I stumbled upon this [doc page](https://docs.microsoft.com/en-us/azure/app-service/app-service-authentication-how-to#configuration-file-reference) which refers to this field as a `<timespan>`. Then I tracked down this [StackOverflow post](https://stackoverflow.com/a/50157340) that describes how .NET `Timespans` are string-formatted.

🤷
